Add documentation pages for metrics and traces (#1686)
google/adk-docs: docs/observability/metrics.md

> # Agent activity metrics

Reviewer: I don't see a mention of `--otel_to_cloud` flag.

I think the doc should be structured like:

```
# Agent activity metrics
## Metrics Philosophy
## Metrics schema
## Metrics export setup
### Metrics export in ADK Web
#### GCP export
#### OTLP export
### Programatic metrics export
#### GCP export setup
#### OTLP export setup
```

<div class="language-support-tag">
  <span class="lst-supported">Supported in ADK</span><span class="lst-python">Python v0.1.0</span><span class="lst-typescript">TypeScript v0.2.0</span><span class="lst-go">Go v0.1.0</span><span class="lst-java">Java v0.1.0</span>
</div>

Agent Development Kit (ADK) provides built-in, vendor-neutral metrics collection to help you understand the performance, cost, and usage patterns of your agents. While logs provide a detailed narrative of *what* happened, metrics give you aggregated, quantitative data to answer *how often* and *how fast* things are happening.

## Metrics Philosophy

ADK's approach to metrics is designed to be lightweight, standardized, and entirely agnostic to your choice of monitoring backend.

> *   **OpenTelemetry Native:** ADK implements the OpenTelemetry (OTel) [Semantic Conventions for GenAI](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-metrics.md). This means your metrics will seamlessly integrate into any OTel-compatible backend (e.g., Prometheus, Datadog, SigNoz, Google Cloud Monitoring).

Reviewer: Semantic Conventions don't have to do with compatibility.

There's the:
- OTLP format - How data is emitted, the actual wire format
- OTel semconv - Under what attribute/span/log/metric names you can expect certain entities to be recorded

I'd split this into two points

> *   **Cost and Performance Focused:** By default, ADK tracks the most critical signals for LLM applications: token consumption, request latency, and tool execution reliability.

Reviewer: The "By default" insinuates you can opt into additional metric export, which IIRC you can't. The point here is valid, but I would compare metrics with logs/traces (i.e. they are less costly, and more performant if you'd like to perform analytics over large swathes of data)

*   **Vendor-Neutral Export:** ADK does not lock you into a specific metrics pipeline. You instantiate standard OTel meter providers and export data wherever your infrastructure demands.

---

## Core Agent Metrics

When metrics are enabled, ADK automatically instruments the agent's lifecycle, LLM interactions, and tool executions based on the OpenTelemetry GenAI Semantic Conventions. The following core metrics are emitted:

> | Metric Name | Type | Description | Key Attributes (Dimensions) |

Reviewer: It makes sense to include metrics from:

https://github.com/google/adk-python/commit/6942aac5d7b1f465c20febe2a48bac90da32c4eb

Into this doc

| :--- | :--- | :--- | :--- |
| **[`gen_ai.client.token.usage`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-metrics.md#metric-gen_aiclienttokenusage)** | Histogram | Tracks the number of input and output tokens consumed per request. Crucial for monitoring costs and context window limits. | `gen_ai.token.type` (input/output), `gen_ai.request.model`, `gen_ai.operation.name` |
| **[`gen_ai.client.operation.duration`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-metrics.md#metric-gen_aiclientoperationduration)** | Histogram | The duration of GenAI operations. This single metric covers end-to-end agent runs, LLM inference requests, and individual tool executions. | `gen_ai.operation.name`, `gen_ai.agent.name`, `gen_ai.request.model`, `error.type` |

> > **Note:** To see the specific duration of an agent invocation versus a tool execution, filter or group the `gen_ai.client.operation.duration` metric by the `gen_ai.operation.name` attribute (e.g., `invoke_agent`, `chat`, `execute_tool`).

Reviewer: I don't think we emit `gen_ai.client.operation.duration` for `invoke_agent` or `execute_tool`?

I think we emit `gen_ai.agent.invocation.duration` instead.

Also we never expect to see `chat` operation name in ADK, we use `generate_content` instead. In fact I'm pretty sure this can **only** be `generate_content` so mentioning filtering likely doesn't make sense

---

## Configuring Metrics with the ADK CLI

If you are running your agent using the `adk run` or `adk web` CLI commands, you can enable OTLP metric exports via environment variables, adhering to the standard OTel specification:
```bash

> export OTEL_EXPORTER_OTLP_METRICS_ENDPOINT="http://your-collector:4318/v1/metrics"

Reviewer: I would breadcrumb here that user can set `OTEL_EXPORTER_OTLP_ENDPOINT`, if they'd like to send traces and logs to the same endpoint in addition to metrics

> export OTEL_METRICS_EXPORTER="otlp"

Reviewer: What does the `OTEL_METRICS_EXPORTER` env do?

What happens if we don't set it, what's the default export format?

Isn't it enough to just set the `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT`?

adk web path/to/your/agents_dir
```

## Configuring Metrics in Python

In Python, ADK metrics are captured by configuring the standard `opentelemetry` Python SDK.

### Example Configuration (OTLP Export)

> To enable metrics and export them to an OpenTelemetry Collector (or an OTLP-compatible backend), set up your `MeterProvider` before initializing your agent:

Reviewer: The export to "OTLP-compatible" backend. Is handled by the section above **if** user uses `adk web`.

This section goes into what would a user have to write to get metrics when integrating ADK into their application (which might not be a http server, like `adk web`).

We have utilities to make this a bit simpler:

```py
from google.adk.telemetry.setup import maybe_set_otel_providers
import os

os.environ["OTEL_EXPORTER_OTLP_METRICS_ENDPOINT"] = "http://your-collector:4318/v1/metrics"
os.environ["OTEL_SERVICE_NAME"] = "your-adk-agent"
os.environ["OTEL_RESOURCE_ATTRIBUTES"] = "key1=value1,key2=value2"
maybe_set_otel_providers()
```

For users who would like GCP export
```py
from google.adk.telemetry.google_cloud import get_gcp_exporters
from google.adk.telemetry.setup import maybe_set_otel_providers
import os

gcp_exporters = google_cloud.get_gcp_exporters(
  enable_cloud_metrics = True,
)
os.environ["OTEL_SERVICE_NAME"] = "your-adk-agent"
os.environ["OTEL_RESOURCE_ATTRIBUTES"] = "key1=value1,key2=value2"
maybe_set_otel_providers(gcp_exporters)
```

# Agent activity metrics

<div class="language-support-tag">

>   <span class="lst-supported">Supported in ADK</span><span class="lst-python">Python v0.1.0</span><span class="lst-typescript">TypeScript v0.2.0</span><span class="lst-go">Go v0.1.0</span><span class="lst-java">Java v0.1.0</span>

Reviewer: Update the support tag to be accurate. Metrics are only in Python, in a recent version

</div>

Agent Development Kit (ADK) provides built-in, vendor-neutral metrics collection to help you understand the performance, cost, and usage patterns of your agents. While logs provide a detailed narrative of *what* happened, metrics give you aggregated, quantitative data to answer *how often* and *how fast* things are happening.

> ## Metrics Philosophy

Reviewer 2: ```suggestion
## Metrics philosophy
```

Change heading to sentence case for consistency with the rest of the page.

When metrics are enabled, ADK automatically instruments the agent's lifecycle, workflow steps, and tool executions based on the OpenTelemetry GenAI Semantic Conventions. The following core metrics are emitted:

| Metric Name | Type | Description | Key Attributes (Dimensions) |
| :--- | :--- | :--- | :--- |
| **`gen_ai.agent.invocation.duration`** | Histogram | The total time taken for an agent to process a prompt and return a response. | `gen_ai.agent.name`, `error.type` |
| **`gen_ai.tool.execution.duration`** | Histogram | The execution latency of individual tools called by the agent. Useful for spotting slow external APIs. | `gen_ai.tool.name`, `error.type` |
| **`gen_ai.agent.request.size`** | Histogram | The size or complexity of the incoming request sent to the agent. | `gen_ai.agent.name` |
| **`gen_ai.agent.response.size`** | Histogram | The size or complexity of the final response generated by the agent. | `gen_ai.agent.name` |
| **`gen_ai.agent.workflow.steps`** | Histogram | Tracks the number of iterative steps or reasoning loops an agent takes to complete a workflow. | `gen_ai.agent.name` |

---

## Metrics export setup

### Metrics export in ADK Web

> If you are running your agent using the `adk run` or `adk web` CLI commands, you can configure metrics export.

Reviewer 2: ```suggestion
If you are running your agent using the `adk web` or `adk api_server` CLI commands, you can configure metrics export.
```

`adk run` does not support `--otel_to_cloud` or OTLP environment variable configuration. The traces page correctly lists adk web and adk api_server; this page should match.

#### GCP export

To enable metrics export to Google Cloud Monitoring, use the `-otel_to_cloud` flag:

```bash
adk web -otel_to_cloud path/to/your/agents_dir
```

### Programmatic metrics export

You can also configure metrics export programmatically in your application code.

#### OTLP export setup

To enable metrics and export them to an OpenTelemetry Collector (or an OTLP-compatible backend) programmatically:

```python
from google.adk.telemetry.setup import maybe_set_otel_providers
import os

os.environ["OTEL_EXPORTER_OTLP_METRICS_ENDPOINT"] = "http://your-collector:4318/v1/metrics"
os.environ["OTEL_SERVICE_NAME"] = "your-adk-agent"
os.environ["OTEL_RESOURCE_ATTRIBUTES"] = "key1=value1,key2=value2"
maybe_set_otel_providers()
```

#### GCP export setup

To export metrics to Google Cloud Monitoring programmatically, use the OpenTelemetry Google Cloud exporter. Here is an example in Python:

```python
from google.adk.telemetry.google_cloud import get_gcp_exporters
from google.adk.telemetry.setup import maybe_set_otel_providers
import os

> gcp_exporters = google_cloud.get_gcp_exporters(

Reviewer 2: ```suggestion
gcp_exporters = get_gcp_exporters(
```

Change to match the scoped import above.

  enable_cloud_metrics = True,
)
os.environ["OTEL_SERVICE_NAME"] = "your-adk-agent"
os.environ["OTEL_RESOURCE_ATTRIBUTES"] = "key1=value1,key2=value2"

> maybe_set_otel_providers(gcp_exporters)

Reviewer 2: ```suggestion
maybe_set_otel_providers([gcp_exporters])
```

`maybe_set_otel_providers` expects `list[OTelHooks]` as its first argument, but this passes a single `OTelHooks` object.
